pkg/services/gc/gc.go

```diff
@@ -0,0 +1,341 @@
+package gc
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"strings"
+
+	"github.com/go-logr/logr"
+	"golang.org/x/exp/maps"
+	authorizationv1 "k8s.io/api/authorization/v1"
+	k8serr "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/discovery"
+	ctrlCli "sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/controller/client"
+)
+
+// Instance a global instance of the GC service.
+//
+// TODO: since the GC service is quite heavy, as it has to discover
+//
+//	resources that can be subject to GC, we share single global
+//	instance, however as long term, we should find a better way
+//	to let consumer of the service to access it.
+var Instance *GC
+
+const (
+	DeleteVerb  = "delete"
+	AnyVerb     = "*"
+	AnyResource = "*"
+)
+
+type options struct {
+	propagationPolicy ctrlCli.PropagationPolicy
+	unremovables      []schema.GroupVersionKind
+}
+
+type OptsFn func(*options)
+
+func WithUnremovables(items ...schema.GroupVersionKind) OptsFn {
+	return func(o *options) {
+		o.unremovables = append(o.unremovables, items...)
+	}
+}
+
+func WithPropagationPolicy(policy metav1.DeletionPropagation) OptsFn {
+	return func(o *options) {
+		o.propagationPolicy = ctrlCli.PropagationPolicy(policy)
+	}
+}
+
+func New(cli *client.Client, ns string, opts ...OptsFn) *GC {
+	res := GC{
+		client: cli,
+		ns:     ns,
+		options: options{
+			propagationPolicy: ctrlCli.PropagationPolicy(metav1.DeletePropagationForeground),
+			unremovables:      make([]schema.GroupVersionKind, 0),
+		},
+
+		resources: Resources{
+			items: make([]Resource, 0),
+		},
+	}
+
+	for _, o := range opts {
+		o(&res.options)
+	}
+
+	return &res
+}
+
+type GC struct {
+	client    *client.Client
+	ns        string
+	resources Resources
+	options   options
+}
+
+func (gc *GC) Start(ctx context.Context) error {
+	l := gc.log(ctx)
+	l.Info("Start computing deletable types")
+
+	res, err := gc.computeDeletableTypes(ctx)
+	if err != nil {
+		return fmt.Errorf("cannot discover deletable types: %w", err)
+	}
+
+	gc.resources.Set(res)
+
+	l.Info("Deletable types computed", "count", gc.resources.Len())
+
+	return nil
+}
+
+func (gc *GC) Run(
+	ctx context.Context,
+	selector labels.Selector,
```

Comment:
we define `selector` and `lables`, all in all, it is to be 
`lo := metav1.ListOptions{LabelSelector: labels.SelectorFromSet(action.labels).String()} ` ?

Reply:
yes possible, however the `labels.SelectorFromSet` only support equality, whereas a plain selector can also support additional selection comparison, hence I did not want to limit the options since the gc service can be used for additional use cases i.e. 

```go
releases := labels.NewRequirement("platform.opendatahub,io", selection.In, []string{"2.14.1", "2.14.2"})


selector := labels.NewSelector()
selector = selector.Add(*releases)
```